### ISSUE TITLE
Syntax error is not showing for signature mismatch

### ISSUE THREAD

**Issue by umstmo** [microsoft/TypeScript#36940]
**TypeScript Version:**  3.7.5

**Type checking syntax error** 

**Code**

```ts
class Hell {
    name?: string;
    constructor(name: string) {
        this.name = name;
    }
    sayHello() {
        console.log(this.name);
    }
}

class Heaven {
    name1?: string;
    constructor(name: string) {
        this.name1 = name;
    }
    sayHello() {
        console.log(this.name1);
    }
}

class God {
    test: Hell;
    // constructor accept the object of the type 'Hell'
    constructor(test: Hell) {
        this.test = test;
    }
    howdie() {
        console.log(this.test.sayHello());
    }
}

let hell = new Hell('helloo');
let heaven = new Heaven('doomed');

// but its not showing type checking error while passing the object of the type 'Heaven' 
let dingan = new God(heaven);


```

**Expected behavior:   It should show syntax error for signature mismatch**

**Actual behavior: No error is thrown and the code compiles without any issues**

**Playground Link: https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=46&pc=1#code/MYGwhgzhAEASCmITQN4ChqegOzAW3gH4AuaCAFwCcBLbAcwG4MtgB7bCygV2HNcoAUuAqU606ASlTMsWcgAtqEAHTD40ALw588JrIC+MsmACeCJKwFT0s2Ww6sQ8ZSFZ0BCpap0S9WQ4ZooJAwCGAAbvDY0rJqAIwkZFTifpj2nDx8gmqiyfTWRrKeKvGa2gSp0IayEKbmrlYxtmnsEI7Oru7F3gRxvkYBaEPBUNAA4qwAJk2Y5PAUpPWVAPTL0OlUmfzQYMDA8AAO5NAK6qwARgBW8LzQrABmJ-Lq5CYH6gDk9R9GG9y8-A883Ii0QIAKzVmihUcwoZVh5Eq1Sw8lYAHdJtR4I0bJD0u0XG4PNDlAjlLUzGDLBJ+gY0IE0E5js8kGVsPA0XAwQIPizXKwPv0mdBnhEomyOVyxdgeZNWKwCJNBUwhqtoOcuMdqOQYNhWMcIKi0eITm91MBnsAANYm+CUSjbNGKJzQA4hE2nO5XG7HB5PF5m6BfeDSj7QRnwY6Y+hgaJadmciaTASiyLYfpDTNoIA**

**Related Issues:**

**Comment by MartinJohns** [comment]
Your classes have no private/protected members and as a result are structurally typed. Both your classes are structurally equal.

https://github.com/microsoft/TypeScript/wiki/FAQ#when-and-why-are-classes-nominal

**Comment by umstmo** [comment]
Even with two different classes having different property names and also with different types it is accepting the wrong class name. Please see the below example for more clarity. 

https://www.typescriptlang.org/play/index.html?ssl=1&ssc=1&pln=46&pc=1#code/MYGwhgzhAEASCmITQN4ChqemA5vA-AFzQB2ArgLYBG8ATgNwZbAD2JEALrWcBy7QApc8YuWp0AlKiZYsHABYBLCADph0ALzY8jWQF8Z0CGACeCJCwFT0s2a3YsQ8FSBY4BC5WrwTdWAwZooJAwCGAAbvAk0rIkYBQExJy0iiQ4fpj2yTx8gnEJSVypONaGsp6q+fCapPHwGdAGssZmiK5WMbaZbBCOzq7uFSpVvoYBaBPBUNAA4iwAJp2YHPCcxOYgDQD0W9BZXDn82MDA8AAOHNAK1SxUAFbwvNAsAGZX8tUcJmfVAOQbv0M+24vH4HlWHHWbVKXWWSlUK04NURHAaTSw8hYAHd5op4B0bLCsn0XG4PPCVCiVC0NpYJKN9GhAmgnJcPkgaiR4Fi4G0BABGABMo1Z0A+ESinO5vIlJAEv3mLBYCXmv1GEx20CoZEuig4MBILEuEExWOKV2+1WAH2AAGtzXRaEcsUonNAziFzddnvdHpdXu9PpboP94LLftAWfBLri0mBolouTy5vMBOLIiR1RNs0A

**Comment by MartinJohns** [comment]
Your types are still structurally compatible. The only required member of `Hell` is `sayHello()`, and that is present within `Heaven`. The property `age` is not considered because it's optional, it may be `undefined ` in `Hell` and in `Heaven` it is `undefined` because it's a property of the class - so it's compatible.

https://github.com/microsoft/TypeScript/wiki/FAQ#what-is-structural-typing
